vortex-array/src/array/constant/canonical.rs

```diff
@@ -1,62 +1,100 @@
 use std::iter;
 
-use vortex_dtype::{match_each_native_ptype, DType, Nullability, PType};
+use vortex_dtype::{match_each_native_ptype, DType, Nullability};
 use vortex_error::{vortex_bail, VortexResult};
-use vortex_scalar::{BinaryScalar, BoolScalar, Utf8Scalar};
+use vortex_scalar::ScalarValue;
 
 use crate::array::constant::ConstantArray;
 use crate::array::primitive::PrimitiveArray;
 use crate::array::varbin::VarBinArray;
-use crate::array::BoolArray;
+use crate::array::{BoolArray, NullArray};
 use crate::validity::Validity;
 use crate::{ArrayDType, Canonical, IntoCanonical};
 
 impl IntoCanonical for ConstantArray {
     fn into_canonical(self) -> VortexResult<Canonical> {
         let validity = match self.dtype().nullability() {
             Nullability::NonNullable => Validity::NonNullable,
-            Nullability::Nullable => match self.scalar().is_null() {
+            Nullability::Nullable => match self.scalar_value().is_null() {
                 true => Validity::AllInvalid,
                 false => Validity::AllValid,
             },
         };
 
-        if let Ok(b) = BoolScalar::try_from(self.scalar()) {
-            return Ok(Canonical::Bool(BoolArray::from_vec(
-                vec![b.value().unwrap_or_default(); self.len()],
-                validity,
-            )));
-        }
-
-        if let Ok(s) = Utf8Scalar::try_from(self.scalar()) {
-            let value = s.value();
-            let const_value = value.as_ref().map(|v| v.as_bytes());
+        let dtype = self.dtype().clone();
 
-            return Ok(Canonical::VarBin(VarBinArray::from_iter(
-                iter::repeat(const_value).take(self.len()),
-                DType::Utf8(validity.nullability()),
-            )));
-        }
+        match self.scalar_value() {
```

Comment:
I think it's probably better to just call `self.owned_scalar()` here and then keep the existing logic so that we don't have to implicitly preserve the `<$Name>Scalar::try_from` behavior in this function